Add hooks for overriding surface radiative fluxes seen by land and ocean models

spencerkclark opened the pull request:
**Description**

This PR adds hooks to enable overriding the surface radiative fluxes seen by the land and ocean models following https://github.com/ai2cm/fv3gfs-fortran/pull/158 in AI2's fork of FV3GFS.  [This document](https://paper.dropbox.com/doc/2021-03-12-design-of-the-fortran-hooks-for-overriding-the-surface-radiative-fluxes-seen-by-the-surface-schemes-o4oxKjCgRxbfjrXMrCj7c) contains a description of the thought process behind this code in FV3GFS, which has been used in several studies.  In short it:
- Adds fields to a new data structure (`IPD_Data%OverridesFromPythonWrapper`) that are meant to hold data for overriding the surface radiative fluxes (namely the downward longwave, downward shortwave, and net shortwave radiative components).
- Creates pointers in the physics driver for these fluxes, which are set to either point to the adjusted values from the radiative transfer code (i.e. how the model normally runs) or the overriding fluxes, depending on the value of the `gfs_physics_nml.override_surface_radiative_fluxes` namelist parameter.
- Configures the radiative flux diagnostics to reflect what the surface felt, and adds additional diagnostics for the radiative fluxes output by the radiative transfer code when they are being overridden (i.e. the `_from_rrtmg` fields).

Note that this PR also must be coupled with a PR to NOAA-GFDL/atmos_drivers to follow through with passing the `override_surface_radiative_fluxes` namelist parameter to the diagnostics initialization routine (see NOAA-GFDL/atmos_drivers#31).

@lharris4 this should be the last set of PRs needed to obtain feature parity with our wrapper of FV3GFS.  Since this does something science-related, I'm guessing it may require a more careful review.

**How Has This Been Tested?**

These changes have been tested in https://github.com/ai2cm/SHiELD-wrapper/pull/5.  There we test that overriding the fluxes changes the solution of the model, and that the diagnostics are updated as expected based on the overridden values.  As expected, the answers for the model also do not change in configurations where `gfs_physics_nml.override_surface_radiative_fluxes` is `.false.` (the default).

**Checklist:**
Please check all whether they apply or not
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

bensonr commented on GFS_layer/GFS_physics_driver.F90:
```diff
@@ -617,6 +619,16 @@ subroutine GFS_physics_driver                         &
       endif
       
+      if (Model%override_surface_radiative_fluxes) then
+        adjsfcdlw_for_lsm => Statein%adjsfcdlw_override
+        adjsfcdsw_for_lsm => Statein%adjsfcdsw_override
+        adjsfcnsw_for_lsm => Statein%adjsfcnsw_override
```
Where is Statein being populated with these override values.  While the ai2cm version of the model may have used Statein for this purpose, I'm not sure that's something we want to proliferate if SHiELD is to have a long life at GFDL.  I'd prefer to see a new type in GFS for data overrides or use the Coupler type which will eventually be used to move data back and forth to a decoupled land model.  I'll let @lharris4 make the call on this though.

spencerkclark replied:
Thanks for taking a look @bensonr.  Indeed this could be made clearer; since the wrapper is being used to populate the override values it will always look a bit mysterious from the fortran perspective.  We basically just need some allocated state in the fortran model that the wrapper can point to, which could be in `Statein`, a new type, or the `Coupler` type.  A new dedicated type (with a comment describing the intent) is maybe best since the use-case is quite different than anything else that exists currently.  Happy to take guidance from you and @lharris4 on this.

lharris4 replied:
I think the coupler type may be the best choice, if it already has these fields defined within it. The coupler will probably be getting heavier use soon as we start working on SHiELD coupling.

lharris4 replied:
BTW is `data_override()` available within the SHiELD Physics? This may be a better route since it is a standard, supported FMS functionality, and we are already using this for the SSTs.

bensonr replied:
@spencerkclark will correct me if I'm wrong, but I believe these fields are coming from the Python-based ML/DA system directly and not via datafiles read and or interpolated.  So in this case, the data_override functionality from FMS would not be applicable here.

spencerkclark replied:
> I think the coupler type may be the best choice, if it already has these fields defined within it. The coupler will probably be getting heavier use soon as we start working on SHiELD coupling.

Thanks @lharris4 this seems OK so long as it will not be confusing for anyone reading the code.  I'll try and make things clear with a comment.

> BTW is `data_override()` available within the SHiELD Physics? This may be a better route since it is a standard, supported FMS functionality, and we are already using this for the SSTs.

Indeed @bensonr beat me to it.  The tricky thing here is that the data we would like to use for overriding may not necessarily be read in from a file (as I understand `data_override` would require).  It instead can be predicted by an ML model as the simulation is running.

lharris4 replied:
I understand better now @bensonr @spencerkclark. This is different from the usual `data_override()` use case. Indeed going through the datatypes will be a better way. This will probably also be useful for various DA routines like what Naomi is setting up in LM4.

lharris4 reviewed (changes requested):
I agree with @bensonr. I think that using the existing `Coupler` type or a new override data type would be cleaner and avoids needing to change the FV3 state.

spencerkclark commented:
Sounds good @lharris4.  I'll pick this back up next week.

lharris4 commented on GFS_layer/GFS_typedefs.F90:
```diff
@@ -1196,6 +1197,11 @@ module GFS_typedefs
     real (kind=kind_phys), pointer :: gflux  (:)    => null()   !< groud conductive heat flux
     real (kind=kind_phys), pointer :: dlwsfc (:)    => null()   !< time accumulated sfc dn lw flux ( w/m**2 )
     real (kind=kind_phys), pointer :: ulwsfc (:)    => null()   !< time accumulated sfc up lw flux ( w/m**2 )
+    real (kind=kind_phys), pointer :: dswsfc (:)     => null()   !< time accumulated sfc dn sw flux ( w/m**2 ) when gfs_physics_nml.override_surface_radiative_fluxes == .true.
```
These new `dswsfc`/`uswsfc` variables: are they intended to replace the `fluxr` array so they hold the true (overridden) surface fluxes? Was this because modifying `fluxr` could have undesirable side effects?

lharris4 commented on GFS_layer/GFS_typedefs.F90:
```diff
@@ -3519,6 +3579,9 @@ subroutine control_print(Model)
       print *, ' kdt               : ', Model%kdt
       print *, ' jdat              : ', Model%jdat
+
+      print *, 'whether to override surface radiative fluxes seen by the land surface model'
```
I don't think this line is really necessary. The (good!) variable name `override_surface_radiative_fluxes` is pretty self-explanatory, and there is already an awful lot of stdout from the physics.

lharris4 reviewed (changes requested):
Thanks so much for putting this together so quickly and doing such thorough testing. I don't see any problems with your code. I just have a couple of minor cosmetic suggestions and a few things I would be interested in making sure I understand correctly (hence 'minor revisions').

bensonr commented on GFS_layer/GFS_typedefs.F90:
```diff
@@ -885,6 +885,7 @@ module GFS_typedefs
     character(len=32)    :: iau_forcing_var(20)  ! list of tracers with IAU forcing
     real(kind=kind_phys) :: iaufhrs(7)      ! forecast hours associated with increment files
     logical :: iau_filter_increments, iau_drymassfixer
+    logical :: override_surface_radiative_fluxes  ! Whether to use Statein to override the surface radiative fluxes
```
With the Coupling type now holding the data, I think the comment needs to be updated.

bensonr commented on FV3GFS/FV3GFS_io.F90:
```diff
@@ -3087,12 +3088,20 @@ subroutine gfdl_diag_register(Time, Sfcprop, Gfs_diag, Cldprop, &
-    do nb = 1,nblks
-      Diag(idx)%data(nb)%var2 => Gfs_diag(nb)%fluxr(:,3)
-    enddo
+    if (override_surface_radiative_fluxes) then
```
Could use a bit more explanation of what's happening here as well as the few blocks below where you are choosing between a named variable and one of the `fluxr` constituents.

spencerkclark replied:
Good idea.  I'll add a comment each place it occurs.

spencerkclark commented on GFS_layer/GFS_typedefs.F90:
```diff
@@ -1317,6 +1327,30 @@ module GFS_typedefs
   end type GFS_diag_type
 
+!----------------------------------------------------------------
+! GFS_overrides_from_python_wrapper_type
```
Sounds good.  It's also nice that is more concise.

spencerkclark commented on FV3GFS/FV3GFS_io.F90:
```diff
@@ -4628,6 +4653,123 @@ subroutine gfdl_diag_register(Time, Sfcprop, Gfs_diag, Cldprop, &
       Diag(idx)%data(nb)%var2 => Gfs_diag(nb)%uswsfci(:)
     enddo
 
+    if (override_surface_radiative_fluxes) then
```
That's correct.  I'll go ahead and switch from ending with "due to RRTMG" to starting with "Native RRTMG."

spencerkclark reviewed (commented):
Thanks @lharris4 and @bensonr for the review.  I'll push some updates shortly.

spencerkclark commented on FV3GFS/FV3GFS_io.F90:
```diff
@@ -4624,9 +4674,133 @@ subroutine gfdl_diag_register(Time, Sfcprop, Gfs_diag, Cldprop, &
+    endif
+
+    if (override_surface_radiative_fluxes) then
+      idx = idx + 1
+      Diag(idx)%axes = 2
+      Diag(idx)%name = 'DLWRF_from_rrtmg'
+      Diag(idx)%desc = 'Native RRTMG surface downward longwave flux'
+      Diag(idx)%unit = 'W/m**2'
+      Diag(idx)%mod_name = 'gfs_phys'
+      Diag(idx)%cnvfac = cn_one
+      Diag(idx)%time_avg = .TRUE.
+      Diag(idx)%intpl_method = 'bilinear'
+      Diag(idx)%coarse_graining_method = 'area_weighted'
+      allocate (Diag(idx)%data(nblks))
+      do nb = 1,nblks
+        Diag(idx)%data(nb)%var2 => Gfs_diag(nb)%dlwsfc(:)
+      enddo
```
@lharris4 do you happen to know the difference between [this version](https://github.com/NOAA-GFDL/SHiELD_physics/blob/69d13c245348264f1bf12fd871261d5e25e36695/FV3GFS/FV3GFS_io.F90#L3112-L3124) of the interval-averaged surface downward longwave flux diagnostic and [this version](https://github.com/NOAA-GFDL/SHiELD_physics/blob/69d13c245348264f1bf12fd871261d5e25e36695/FV3GFS/FV3GFS_io.F90#L3938-L3950)?  Something similar can be asked for `ULWRFsfc` vs. `ULWRF`.  

It looks like in the PIRE simulations `DLWRFsfc` and `ULWRFsfc` were used.  Are we to assume those are the proper versions of these diagnostics, and that the `DLWRF` and `ULWRF` diagnostics are not to be used?  The reason for asking is that it influences how I would define these RRTMG fields — currently they are defined like the `DLWRF` and `ULWRF` versions, which were the only versions present in FV3GFS.

spencerkclark replied:
Per our offline conversation I went ahead and used `DLWRFsfc` and `ULWRFsfc` as the source of truth for these RRTMG fields.

lharris4 replied:
Actually I am not entirely sure what the different diagnostics represent. It looks like `DLWRF` etc. use a [zenith-angle adjustment](https://github.com/NOAA-GFDL/SHiELD_physics/blob/69d13c245348264f1bf12fd871261d5e25e36695/GFS_layer/GFS_physics_driver.F90#L868-L884) whereas `DLWRFsfc` [is straight from RRTMG](https://github.com/NOAA-GFDL/SHiELD_physics/blob/69d13c245348264f1bf12fd871261d5e25e36695/gsmphys/radsw_main.f#L61-L88) and does not include this correction.

lharris4 replied:
@linjiongzhou @MingjingTong-NOAA do either of you know more about the difference between `DLWRF` and `DLWRFsfc`? This may explain why we see the four 'peaks' in our radiation diagnostics.

spencerkclark replied:
That's an important point @lharris4, thanks.  In the case of the longwave fluxes, those are "surface temperature adjusted" and indeed those are the ones that we should actually use here since that is in fact what we are overriding (despite what we may have output in past simulations).  I can fix that tomorrow morning.

In the case of shortwave it looks like [there are instantaneous diagnostics for the zenith-angle-adjusted versions](https://github.com/NOAA-GFDL/SHiELD_physics/blob/69d13c245348264f1bf12fd871261d5e25e36695/FV3GFS/FV3GFS_io.F90#L4607-L4629) but not interval-averaged.  I can add those as well, as I think they could be useful going forward (those are again what we are actually overriding).

MingjingTong-NOAA replied:
I took a look at the code. It seems DLWRFsfc directly comes out of radiation calculation in subroutine GFS_radiation_driver, while DLWRF is time step adjusted sfc downward lw flux coming out from subroutine dcyc2t3.

!    dcyc2t3 fits radiative fluxes and heating rates from a coarse      !
!    radiation calc time interval into model's more frequent time steps.!

linjiongzhou replied:
Based on my understanding, Xsfc is the averaged radiation flux within the radiation time step (e.g., 1 hour), Xsfci is the averaged radiation flux within the physics time step (e.g., 150s), Xsfci is called instantaneous flux probably due to its short time interval. Xsfci is adjusted from Xsfc based on the difference of solar zenith angle between the radiation and physics time steps. The shape of Xsfci should be taller and narrower than the Xsfc. As for X, it is Xsfci * dt, which should hold a different unit of J/m^2. It is a time-accumulated variable.

spencerkclark replied:
Per https://github.com/NOAA-GFDL/SHiELD_physics/pull/31#discussion_r1373903230, in 176da07 I went ahead and:

- Switched the toggling between RRTMG and overridden fluxes to occur only for the diagnostics that reflect the "adjusted" flux, i.e. `DLWRF`, `DLWRFI`, `DSWRF`, `DSWRFI`, `USWRF`, and `USWRFI`, since that is ultimately what we are overriding.  
    - Note that `DSWRF` and `USWRF` are newly added diagnostics that did not exist previously.
- I took the liberty to add more detail to the `desc` fields of all of the surface radiative flux diagnostics, to make clear the distinction between fields that come unadjusted from the radiative transfer code, and fields that are adjusted.

@lharris4 @MingjingTong-NOAA @linjiongzhou let me know if you agree with these changes.  

My read of the "instantaneous" description is that it is to distinguish from "interval-averaged" fields ("interval-average" meaning fields for which `Diag(idx)%time_avg = .TRUE.`).  The interval averages [are computed by dividing by the length of the averaging interval](https://github.com/NOAA-GFDL/SHiELD_physics/blob/69d13c245348264f1bf12fd871261d5e25e36695/FV3GFS/FV3GFS_io.F90#L7126), so the accumulated time-weighted fluxes of `Diag%dlwsfc` et al. are ultimately transformed back to having units of W/m^2.

linjiongzhou replied:
@spencerkclark, you are right. They are eventually transformed back to the unit of W/m^2 in FV3GFS_io.F90, for both Xsfc and X. Thanks for digging into this and clarifying it.

spencerkclark commented:
I think this should be ready for re-review.  Spurred on by https://github.com/NOAA-GFDL/SHiELD_physics/pull/31#discussion_r1370758153 and https://github.com/NOAA-GFDL/SHiELD_physics/pull/31#discussion_r1371003375 I have refactored the diagnostics related code to always use the existing diagnostics buckets (`fluxr`, `dlwsfci` and the like) for RRTMG-related data, and always use newly added `_override` buckets for data from the overrides.  This should hopefully make things a little more readable than before, where I was sometimes using existing buckets for overrides and sometimes not.  I have addressed the other comments following my inline responses above.

lharris4 commented on FV3GFS/FV3GFS_io.F90:
```diff
@@ -3082,7 +3083,7 @@ subroutine gfdl_diag_register(Time, Sfcprop, Gfs_diag, Cldprop, &
     Diag(idx)%axes = 2
     Diag(idx)%name = 'USWRFsfc'
-    Diag(idx)%desc = 'averaged surface upward shortwave flux'
+    Diag(idx)%desc = 'Interval-averaged unadjusted upward shortwave flux at the surface'
```
I really appreciate these clarifications. Thanks.

lharris4 reviewed (approved):
Thanks for putting in the extra effort to answer my many questions. I think this is good to go.

spencerkclark commented:
Excellent, thanks @lharris4!  Should we be good to merge this and the related https://github.com/NOAA-GFDL/atmos_drivers/pull/31 now?

lharris4 commented:
Glad to help :-) I think it should be good but I will defer to Rusty's team
for the final word on the driver modifications.

On Wed, Nov 1, 2023 at 8:51 AM Spencer Clark ***@***.***>
wrote:

> Excellent, thanks @lharris4 <https://github.com/lharris4>! Should we be
> good to merge this and the related NOAA-GFDL/atmos_drivers#31
> <https://github.com/NOAA-GFDL/atmos_drivers/pull/31> now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/SHiELD_physics/pull/31#issuecomment-1788903952>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMUQRVF4J6K2CMTGN5TMF5TYCJAWRAVCNFSM6AAAAAA6I7XSMGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOBYHEYDGOJVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
